core/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourceSubPhase.java

```diff
@@ -35,29 +42,50 @@ public void hitExecute(SearchContext context, HitContext hitContext) {
         if (context.sourceRequested() == false) {
             return;
         }
+        boolean rootHit = hitContext.hit().getNestedIdentity() == null;
         SourceLookup source = context.lookup().source();
         FetchSourceContext fetchSourceContext = context.fetchSourceContext();
         assert fetchSourceContext.fetchSource();
         if (fetchSourceContext.includes().length == 0 && fetchSourceContext.excludes().length == 0) {
-            hitContext.hit().sourceRef(source.internalSourceRef());
+            if (rootHit ) {
+                hitContext.hit().sourceRef(source.internalSourceRef());
+            } else {
+                try {
+                    applyNestedSource(source.source(), source.sourceContentType(), hitContext);
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
             return;
         }
 
-        if (source.internalSourceRef() == null) {
+        if (rootHit && source.internalSourceRef() == null) {
```

Comment:
Do we need to check `source.internalSourceRef() == null` before returning to ensure that for non-root hits its actually been set by the end of this method?

Reply:
Yes, I think so. In case of regular hits the source has already been set in the FetchPhase and I guess this check exists to ensure that the binary representation of the _source has really been set.